src/sql/src/query_model/hir/hir_from_qgm.rs

```diff
@@ -24,23 +31,393 @@ impl From<Model> for HirRelationExpr {
+        } else {
+            self.converted.push(hir);
+        }
+    }
+
+    /// Convert the quantifiers of `box` into [HirRelationExpr] and [HirScalarExpr].
+    ///
+    /// Results are returned in reverse quantifier order; the last entry in the
+    /// second vector corresponds to the first subquery quantifier in the box.
+    fn convert_quantifiers<'a>(
+        &mut self,
+        r#box: &BoundRef<'a, QueryBox>,
+    ) -> (
+        Vec<(QuantifierId, HirRelationExpr)>,
+        Vec<(QuantifierId, HirScalarExpr)>,
+    ) {
+        let mut rels = Vec::new();
+        let mut scalars = Vec::new();
+        for q in r#box.input_quantifiers().rev() {
+            // Retrive the HIR corresponding to the input box of the quantifier
+            let inner_relation = if let Some(let_pos) = self.common_subgraphs.get(&q.input_box) {
+                if matches!(self.lets[*let_pos].1, HirRelationExpr::Get { .. }) {
+                    self.lets[*let_pos].1.clone()
+                } else {
+                    HirRelationExpr::Get {
+                        id: mz_expr::Id::Local(self.lets[*let_pos].0),
+                        typ: self.lets[*let_pos].2.clone(),
+                    }
+                }
+            } else {
+                self.converted.pop().unwrap()
+            };
+
+            match q.quantifier_type {
+                QuantifierType::Foreach | QuantifierType::PreservedForeach => {
+                    rels.push((q.id, inner_relation))
+                }
+                QuantifierType::Existential => {
+                    scalars.push((q.id, HirScalarExpr::Exists(Box::new(inner_relation))))
+                }
+                QuantifierType::Scalar => {
+                    scalars.push((q.id, HirScalarExpr::Select(Box::new(inner_relation))))
+                }
+                QuantifierType::All => {
+                    // TODO: Shouldn't QuantifierType::Any also exist?
+                    unimplemented!()
+                }
+            }
+        }
+        (rels, scalars)
+    }
+
+    /// Convert the projection of a box into a Map + Project around the
+    /// [HirRelationExpr] representing the rest of the box.
+    fn convert_projection<'a>(
+        &mut self,
+        r#box: &BoundRef<'a, QueryBox>,
+        column_map: &ColumnMap,
+        mut result: HirRelationExpr,
+    ) -> HirRelationExpr {
+        let mut map = Vec::new();
+        let mut project = Vec::new();
+        let mut total_columns = column_map.arity();
+
+        for column in &r#box.columns {
+            let hir_scalar = self.convert_scalar(&column.expr, column_map);
+            if let HirScalarExpr::Column(ColumnRef { level: 0, column }) = hir_scalar {
+                // We can skip making a copy of the column.
+                project.push(column);
+            } else {
+                map.push(hir_scalar);
+                project.push(total_columns);
+                total_columns += 1;
```

Comment:
Using `project.len()` would break correctness. 

Example query: `select col5 + 1 from five_column_input`

The correct HIR is
```
Get five_column_input
Map #4 + 1
Project #5
```
if we were to use project.len(), the `Project #5` would turn into a `Project #0`.

However, I can use `column_map.arity() + map.len()` without breaking correctness.